doc/reference/std/misc/atom.md

```diff
@@ -0,0 +1,339 @@
+# Clojure-style Atoms
+::: tip To use the bindings from this module:
+``` scheme
+(import :std/misc/atom)
+```
+:::
+
+Atoms provide a way to manage shared, synchronous, independent state,
+closely following the API defined by [Clojure](https://clojure.org/reference/atoms).
+
+Compared to the Clojure atom API, we:
+  - We prefix all function names with `atom-`.
+  - Do not support the meta field and the functions `alter-meta!` or `reset-meta!`
+  - We replace Clojure `swap-vals!` and `reset-vals!` by our `swap-values!` and `reset-values!`
+    that return values rather than vectors.
+  - We `add-watch` and `remove-watch` to `atom-add-watch!` and `atom-remove-watch!`
+    with the `atom-` prefix and the `!` suffix.
+  - Do not have generic functions that work on multiple kinds of "references"
+    but only one set of functions for atoms.
+  - We also export `atom-increment!` and `atomic-counter` for the common case.
+
+You create an atom with `atom`, and can access its state with `atom-deref`.
+Just like Clojure atoms and other references, our atoms support validators:
+functions that validate new values.
+To change the value of an atom, you can use `atom-swap!`,
+that atomically applies a function to the old value,
+and binds the atom to the new value to it the validator accepts it.
+A lower-level `compare-and-set!` is also provided.
+Changes to atoms are always free of race conditions.
+
+Unlike the Clojure implementation, our current implementation does not retry,
+though we do not guarantee not to change in the future to match Clojure.
+Unlike in Clojure, it is thus acceptable for the change function to have side-effects,
+though it is not recommended, in case this behavior changes.
+
+## atom
+``` scheme
+(atom (atom (initial-value (void)) validator: (validator #f))) -> Atom
+```
+
+Creates a new atom, a box that atomically changes value.
+If no initial-value is provided, `(void)` is used (Clojure has no default initial-value).
+A `validator:` keyword argument can specify a validator
+(defaults to `#f` which designates a validator that always says "yes").
+The validator takes a new value and returns true if it is accepted,
+returns `#f` or throws if it is rejected.
+
+::: tip Examples:
+``` scheme
+> (import :std/iter :std/misc/atom)
+> (def a (atom 0))
+> (def workers (for/collect (i (in-range 100)) (spawn atom-swap! a 1+)))
+> (begin (for-each thread-join! workers) (atom-deref a))
+100
+```
+:::
+
+## atom?
+``` scheme
+(atom? a) -> boolean
+```
+
+Returns `#t` if *a* an atom, `#f` otherwise.
+
+::: tip Examples:
+``` scheme
+> (atom? (atom))
+t
+
+> (atom 42)
+#<Atom #8>
+> (atom? #8)
+#t
+```
+:::
+
+## atom-swap!
+``` scheme
+(atom-swap! atom function) -> any
+```
+
+Applies the function to the value currently in the atom, and,
```

Comment:
(Make sure my interpretation is right!)

Applies `function` to the value currently in `atom`. If it passes the validator, sets `atom` to the new value, otherwise `atom` stays unchanged. Ensures that no other access to `atom` is made in the middle of a swap. Returns the (possibly updated) value of `atom`.

Reply:
What's returned if it doesn't validate?